stablehlo/tests/CheckOps.cpp

```diff
@@ -156,6 +157,96 @@ llvm::Error evalExpectSerializedEqOp(const Tensor &expected, StringRef probeId,
   return evalExpectEqOp(expected, *tensor);
 }
 
+static uint64_t ULPDifference(APFloat f, APFloat g) {
+  // z is ULP-distance between exact 0 and largest subnormal:
+  uint64_t z = APFloat::getSmallestNormalized(f.getSemantics())
+                   .bitcastToAPInt()
+                   .getLimitedValue() -
+               1;
+  // a is ULP-distance between exact 0 and abs(f):
+  uint64_t a = (f.isNegative() ? -f : f).bitcastToAPInt().getLimitedValue();
+  // b is ULP-distance between exact 0 and abs(g):
+  uint64_t b = (g.isNegative() ? -g : g).bitcastToAPInt().getLimitedValue();
+  if (f.isFinite() && g.isFinite()) {
+    // subtract subnormals contribution, round subnormals to closest normal or
+    // zero:
+    if (z < a)
+      a -= z;
+    else
+      a = (2 * a < z ? 0 : 1);
+    if (z < b)
+      b -= z;
+    else
+      b = (2 * b < z ? 0 : 1);
+    if (f.isNegative() != g.isNegative()) {
+      // in case of overflow, return uint64 maximal value
+      if (a > std::numeric_limits<uint64_t>::max() - b)
+        return std::numeric_limits<uint64_t>::max();
+      return a + b;
+    }
+  }
+  return (a > b ? a - b : b - a);
```

Comment:
This result is somewhat meaningless if `a` and `b` are not finite. It also treats NaN values with different payloads as different.

Either comment something like "Note this returns an arbitrary nonzero value for non-equal values when one value is nonfinite. NaNs with different payloads are not considered equal". Or, return -1 and handle this in evalExpectIsCloseToReferenceOp

Reply:
ULPDifference implementation is now revised: all NaNs are considered equal because reference functions typically return qNaN (per definition of special values handling, say, in C++ standard) while function implementations may return signaling NaN values depending on the used algorithm.